events/search_index/postgres.py

```diff
@@ -0,0 +1,104 @@
+import logging
+from typing import Optional
+
+from django.contrib.postgres.search import SearchVector
+from django.db.models import OuterRef, Subquery
+from django.utils import timezone
+
+from events.models import Event, EventSearchIndex
+from events.utils import get_field_attr, split_word_bases
+
+logger = logging.getLogger(__name__)
+
+
+class EventSearchIndexService:
+    """
+    Service class for managing the event search index.
+    """
+
+    @classmethod
+    def update_search_index(cls, event: Optional[Event] = None) -> None:
+        """
+        Update the search index for this event.
+        The search index is updated for each language separately.
+        Also updates the search vectors for the event.
+        """
+
+        for lang in ["fi", "sv", "en"]:
+            words = set()
+            for column in event.get_words_columns(lang):
+                row_content = get_field_attr(event, column)
+                if row_content:
+                    split_word_bases(row_content, words, lang)
+
+            for keyword in event.keywords.values_list("name_%s" % lang, flat=True):
+                split_word_bases(keyword, words, lang)
+
+            logger.info(f"Updating search index for {event.id}, words: {words}")
+
+            EventSearchIndex.objects.update_or_create(
+                event=event,
+                defaults={
+                    "place": event.location if event else None,
+                    "event_last_modified_time": event.last_modified_time
+                    if event
+                    else timezone.now(),
+                    "place_last_modified_time": event.location.last_modified_time
+                    if event and event.location
+                    else timezone.now(),
+                    "words_%s" % lang: list(words),
+                },
+            )
+        # Update search vectors for the event
+        cls.update_index_search_vectors(event)
+
+    @classmethod
+    def update_index_search_vectors(cls, event: Optional[Event] = None) -> None:
+        """
+        Update search vectors for the search index.
+
+        If event is given, only update that event's search vector.
+        Otherwise, update all events' search vectors.
+        """
+        if event:
+            qs = EventSearchIndex.objects.filter(event=event)
+        else:
+            qs = EventSearchIndex.objects.all()
+
+        eqs = Event.objects.filter(full_text=OuterRef("pk"))
```

Comment:
As discussed earlier we could lessen the amount of the events to be indexed to future ones samely as done in [x_ongoing filter](https://github.com/City-of-Helsinki/linkedevents/blob/595da7aa6ab6e5b4f6548c534aa482c41253e43b/events/filters.py#L447)